libnetwork/iptables/firewalld.go

```diff
@@ -163,9 +168,15 @@ func checkRunning() bool {
 	if connection == nil {
 		return false
 	}
-	var zone string
-	err := connection.sysObj.Call(dbusInterface+".getDefaultZone", 0).Store(&zone)
-	return err == nil
+	var currentState dbus.Variant
+	currentState, err := connection.sysObj.GetProperty(dbusInterface + ".state")
+
+	var running bool
+	if err == nil {
+		running, _ = regexp.MatchString(".*RUNNING.*", currentState.Value().(string))
```

Comment:
The firewalld.dbus man page seems to indicate that the property value will be one of an enumerated set of constants. Why can't a simple string-equality check be used here? Please document why in a code comment.

If a string-equality check is insufficient, regular expressions are massive overkill for testing if a string contains a fixed substring. [`strings.Contains`](https://pkg.go.dev/strings#Contains) would be the preferred choice.

Reply:
I made that decision based on several issues and bugs filed.  I dont know godbus really well but the interface has changed, the output has changed before and people have found missing support for different output.  I don't want to make the assumption that the output will be consistent.  In addition some developers expect the output should be Marshalled into json and  others expect it to be string output.  There is also no clear documentation regarding what the exact output should be and it does not match the dbus XML schemas I glanced at or the firewalld documentation.  Its not clear what output I should be expecting other than a string containing the word.  Two examples below but there were more.

https://github.com/godbus/dbus/issues/306
https://github.com/godbus/dbus/issues/328

Reply:
The firewalld documentation is fairly clear once you understand the [D-Bus type system](https://pkg.go.dev/github.com/godbus/dbus/v5@v5.1.0#hdr-Conversion_Rules:~:text=See%20the%20official%20specification%20at%20https%3A//dbus.freedesktop.org/doc/dbus%2Dspecification.html%23type%2Dsystem%20for%20more%20information%20on%20the%20D%2DBus%20type%20system.) and [the Property interface](https://github.com/moby/moby/issues/46658#:~:text=the%20standard%20D%2DBus%20property%20interface.) The `state` property has a D-Bus type signature of `s` and is read-only.

> Its not clear what output I should be expecting other than a string containing the enumerated words from the Firewalld documentation.

What else are you expecting? Have you tried logging the value received to see what it actually is in practice?

> [godbus/dbus#306](https://github.com/godbus/dbus/issues/306)

How is this relevant? Where is `json.Marshal(currentState)` being called?

> [godbus/dbus#328](https://github.com/godbus/dbus/issues/328)

How is this relevant? Where is `dbus.MakeVariant` being called?

Reply:
I have logged the value and it is consistent.  That's not the point as far as I know, the point is godbus does not have clear documentation on expected output and the expected output is not what is in the dbus documentation.  If i remember correctly the output is the string '(Extra = RUNNING)'

Reply:
The expected output is not to be found in the godbus documentation. That would be like looking in the documentation of `net/http` for examples of how to use the GitHub REST API. You need to look at [the firewalld D-Bus API docs](https://firewalld.org/documentation/man-pages/firewalld.dbus.html#:~:text=state%20%2D%20s%20%2D%20\(ro,experienced%20a%20failure.) for such documentation.

Reply:
> and the expected input/output is not what is in the dbus/firewalld documentation. If i remember correctly the output is the string '(Extra = RUNNING)'.

The documentation not matching up with observed behavior is an extraordinary claim, and extraordinary claims require evidence. Please provide evidence; ideally some that I could reproduce.

Reply:
I cannot reproduce the behavior consistently but here is an example. I made the change anyway and we'll stick with the documentation.

<img width="1254" alt="Screenshot 2024-06-21 at 4 03 36 PM" src="https://github.com/moby/moby/assets/347096/c626992f-7762-42c6-90c9-e7e4db415f8a">

<img width="1150" alt="Screenshot 2024-06-21 at 4 26 57 PM" src="https://github.com/moby/moby/assets/347096/49e48629-2653-40fa-8899-a8d2564b5230">

Reply:
How did you generate the console output in those screenshots? I don't see anywhere in this PR which would produce anything like that.

Reply:
[I think I solved the mystery.](https://go.dev/play/p/3ucBc3NjWdM) `%!(EXTRA type=value)` [is how package `fmt` signifies a format-string error.](https://pkg.go.dev/fmt#hdr-Format_errors)